### PR TITLE
Pipe: Fix async client returning self too early when downgrade from handshake v2 to v1

### PR DESCRIPTION
**Pull request by luoluoyuyu**
As title.

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/client/IoTDBDataNodeAsyncClientManager.java [hunk 1]
```diff
@@ -233,7 +234,7 @@ public void onError(Exception e) {
           callback);
       waitHandshakeFinished(isHandshakeFinished);
     }
-
+    client.setShouldReturnSelf(true);
     if (exception.get() != null) {
       throw new PipeConnectionException("Failed to handshake.", exception.get());
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/client/IoTDBDataNodeAsyncClientManager.java:239]: If there are exceptions, maybe we should return the client, or else it's leaked and cannot be borrowed again...

**Review by Caideyipi — APPROVED** [review]
LGTM

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/client/IoTDBDataNodeAsyncClientManager.java [hunk 2]
```diff
@@ -199,43 +199,48 @@ public void onError(Exception e) {
...
+        exception.set(null);
+
+        client.setTimeoutDynamically(PipeConfig.getInstance().getPipeConnectorHandshakeTimeoutMs());
+        client.pipeTransfer(
+            PipeTransferDataNodeHandshakeV1Req.toTPipeTransferReq(
+                CommonDescriptor.getInstance().getConfig().getTimestampPrecision()),
+            callback);
+        waitHandshakeFinished(isHandshakeFinished);
+      }
+      if (exception.get() != null) {
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/client/IoTDBDataNodeAsyncClientManager.java:238]:
```suggestion
      }
      
      if (exception.get() != null) {
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/client/IoTDBDataNodeAsyncClientManager.java [hunk 3]
```diff
@@ -199,43 +199,48 @@ public void onError(Exception e) {
...
-          targetNodeUrl.getPort());
-
-      supportModsIfIsDataNodeReceiver = false;
-      isHandshakeFinished.set(false);
-      resp.set(null);
-      exception.set(null);
+    try {
+      client.setShouldReturnSelf(false);
+      // Try to handshake by PipeTransferHandshakeV2Req.
+      final HashMap<String, String> params = new HashMap<>();
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/client/IoTDBDataNodeAsyncClientManager.java:205]:
```suggestion
      client.setShouldReturnSelf(false);
      
      // Try to handshake by PipeTransferHandshakeV2Req.
      final HashMap<String, String> params = new HashMap<>();
```

**Review by SteveYurongSu — APPROVED** [review]

**Pull request merged**